contracts/RolesRegistry/SftRolesRegistry.sol

```diff
@@ -0,0 +1,241 @@
+        );
+        _;
+    }
+
+    /** External Functions **/
+
+    function grantRoleFrom(
+        RoleAssignment calldata _grantRoleData
+    )
+        external
+        override
+        validGrantRoleData(_grantRoleData.nonce, _grantRoleData.expirationDate, _grantRoleData.tokenAmount)
+        onlyOwnerOrApproved(_grantRoleData.grantor, _grantRoleData.tokenAddress)
+    {
+        if (deposits[_grantRoleData.nonce].grantor == address(0)) {
+            // transfer tokens
+            _deposit(_grantRoleData);
+        } else {
+            // nonce exists
+            require(
+                deposits[_grantRoleData.nonce].grantor == _grantRoleData.grantor,
+                'SftRolesRegistry: grantor mismatch'
+            );
+            require(
+                deposits[_grantRoleData.nonce].tokenAddress == _grantRoleData.tokenAddress,
+                'SftRolesRegistry: tokenAddress mismatch'
+            );
+            require(
+                deposits[_grantRoleData.nonce].tokenId == _grantRoleData.tokenId,
+                'SftRolesRegistry: tokenId mismatch'
+            );
+            require(
+                deposits[_grantRoleData.nonce].tokenAmount == _grantRoleData.tokenAmount,
+                'SftRolesRegistry: tokenAmount mismatch'
+            );
+
+            RoleData storage _roleData = roleAssignments[_grantRoleData.nonce];
+            require(
+                _roleData.expirationDate < block.timestamp || _roleData.revocable,
+                'SftRolesRegistry: nonce is not expired or is not revocable'
+            );
+        }
+        _grantOrUpdateRole(_grantRoleData);
+    }
+
+    function _grantOrUpdateRole(RoleAssignment calldata _grantRoleData) internal {
+        roleAssignments[_grantRoleData.nonce] = RoleData(
+            _grantRoleData.grantee,
+            _grantRoleData.expirationDate,
+            _grantRoleData.revocable,
+            _grantRoleData.data
+        );
+
+        emit RoleGranted(
+            _grantRoleData.nonce,
+            UNIQUE_ROLE,
+            _grantRoleData.tokenAddress,
+            _grantRoleData.tokenId,
+            _grantRoleData.tokenAmount,
+            _grantRoleData.grantor,
+            _grantRoleData.grantee,
+            _grantRoleData.expirationDate,
+            _grantRoleData.revocable,
+            _grantRoleData.data
+        );
+    }
+
+    function _deposit(RoleAssignment memory _grantRoleData) internal {
+        deposits[_grantRoleData.nonce] = DepositInfo(
+            _grantRoleData.grantor,
+            _grantRoleData.tokenAddress,
+            _grantRoleData.tokenId,
+            _grantRoleData.tokenAmount
+        );
+
+        _transferFrom(
+            _grantRoleData.grantor,
+            address(this),
+            _grantRoleData.tokenAddress,
+            _grantRoleData.tokenId,
+            _grantRoleData.tokenAmount
+        );
+    }
+
+    function revokeRoleFrom(uint256 _nonce, bytes32 _role, address _grantee) external override {
```

Comment:
Missing check to see if role is `UNIQUE_ROLE`

Reply:
it does not need to, the _role field is not being used